drivers/hmis/app/models/hmis/auth_policies/context_loaders/ce_referral_assignment_loader.rb

```diff
@@ -25,6 +25,57 @@ def initialize(user)
 
     private
 
+    # Returns workflow instance IDs where there are completed steps assigned to
+    # swimlanes that the current user participates in.
+    #
+    # This allows users to see referrals they're involved with through swimlane
+    # participation, even if no steps are currently directly assigned to them.
+    #
+    # Note: this logic is mirrored by `Referral#with_completed_steps_assigned_to_swimlane`
```

Comment:
Instead of duplicating the logic, could `drivers/hmis/app/models/hmis/ce/referral.rb` have a class method that takes a user (you currently have a scope, so that's probably fine, but it might be cleaner to have some smaller methods that return the various bits you need.

Then this could just call the class methods to build up the various bit you need here to prevent duplication of logic.

Reply:
Thanks @eanders. After coming back to this with fresh eyes, I removed the duplication. Instead of breaking it up into smaller methods, I kept one shared method as a class method on the referral class. I like the idea of breaking it up, but since it uses such specific grouping of records it felt more confusing to have those defined in different places.